lstchain/io/event_selection.py

```diff
@@ -65,12 +78,101 @@ class DL3FixedCuts(Component):
     ).tag(config=True)
 
     def gh_cut(self, data):
+        """
+        Applying a fixed global gammaness cut on a given data
+        """
         return data[data["gh_score"] > self.fixed_gh_cut]
 
+    def optimize_gh_cuts(
+        self, data, energy_bins, min_value=0.1,
+        max_value=0.99, smoothing=None, min_events=10
+    ):
+        """
+        Evaluating an optimized energy-dependent gammaness cuts, in a given
+        data, with provided reco energy bins, and other parameters to
+        pass to the pyirf.cuts.calculate_percentile_cut function
+        """
+        #To be sure to have the efficiency in between 0 and 100
+        if self.fixed_gh_efficiency < 1:
+            self.fixed_gh_efficiency *= 100
+
+        gh_cuts = calculate_percentile_cut(
+            data["gh_score"],
+            data["reco_energy"],
+            bins = energy_bins,
+            min_value = min_value,
+            max_value = max_value,
+            fill_value = data["gh_score"].max(),
+            percentile = 100 - self.fixed_gh_efficiency,
+            smoothing = smoothing,
+            min_events = min_events,
+        )
+        return gh_cuts
+
+    def apply_optimized_gh_cuts(self, data, gh_cuts):
+        """
+        Apply a given energy-dependent gh cuts to a data file, along the reco
+        energy bins provided.
+        """
+
+        data["selected_gh"] = evaluate_binned_cut(
+            data["gh_score"],
+            data["reco_energy"],
+            gh_cuts,
+            operator.ge,
+        )
+        return data[data["selected_gh"]]
+
     def theta_cut(self, data):
+        """
+        Apply a fixed global theta cut on a given data
+        """
         return data[data["theta"].to_value(u.deg) < self.fixed_theta_cut]
 
+    def optimize_theta_cuts(
+        self, data, energy_bins, min_value=0.05 * u.deg,
+        fill_value=0.32 * u.deg, max_value=0.32 * u.deg,
+        smoothing=None, min_events=10
+    ):
+        """
+        Evaluating an optimized energy-dependent theta cuts, in a given
+        data, with provided reco energy bins, and other parameters to
+        pass to the pyirf.cuts.calculate_percentile_cut function
```

Comment:
Note that the very fine binning of the theta cuts used by event display comes from an interpolation. The actual calculation happens in much coarser bins. Using to fine binning will result in too un-smooth cuts.

Reply:
Ok. I will add the final comment to the description